### ISSUE TITLE
Art: Modular forest

### ISSUE THREAD

**Issue by madjin** [webaverse/app#1313]
Dungeon / lost woods inspired modular forest. It will be used for generating a maze-like forest one could get lost in.

## Entrance
The forest is to be surrounded by something to see it from afar. This can be a wall with an entrance. Perhaps there's a danger sign at the entrance?

![image](https://user-images.githubusercontent.com/32600939/122082728-3b7df680-cdce-11eb-8e9e-8a7c9eda293e.png)

![brave!UNITO-UNDERSCORE!VAl5BPxR6g](https://user-images.githubusercontent.com/32600939/122078913-f0161900-cdca-11eb-9e57-33603b1bc5c5.jpg)

![brave!UNITO-UNDERSCORE!X0Q8oNfniz](https://user-images.githubusercontent.com/32600939/122081896-82b7b780-cdcd-11eb-9525-8264140dab1d.jpg)

## Modular pieces
Multiples of chunk sizes, size as square / long shape, with variations of entrances and exits. Think like plumbing, some can be L shaped entrance / exit. Keep all these flat.

Regular 20x20 size

![brave!UNITO-UNDERSCORE!AYiZh43EhU](https://user-images.githubusercontent.com/32600939/122081855-79c6e600-cdcd-11eb-926f-4de354e7ea5a.jpg)

Long example

![brave!UNITO-UNDERSCORE!RGzsFuuCq5](https://user-images.githubusercontent.com/32600939/122081864-7c294000-cdcd-11eb-9b48-0490c03c67c1.jpg)

Couple bigger areas of land, expand it out.

## Props
* Rocks
* Signs
* Fences
* Ferns

**Comment by GaladWarder** [comment]
Dream made these/posted them in the Discord earlier - just putting them here for posterity

![DREAM](https://user-images.githubusercontent.com/29031531/122213635-bd0f7a80-cedb-11eb-8085-0fc72c6c6e79.png)

**Comment by avaer** [comment]
The alignment and pieces are great, I just think we need to get the scale right. At the moment it feels large, like a forest for ants.


https://user-images.githubusercontent.com/6926057/122232418-3762df00-ce89-11eb-98b1-446af18001f8.mp4

**Comment by avaer** [comment]
The original references were built for PC's that are like half the size of a full VRM so making the VRM match in terms of screenshots results in geometry that's 2x too big or so.

**Comment by avaer** [comment]
https://www.zeldadungeon.net/wiki/images/2/21/Lost-Woods-Theme.jpg

**Comment by madjin** [comment]
![image](https://user-images.githubusercontent.com/32600939/122567623-b4af6080-cffd-11eb-9eaa-2bcf57104625.png)

**Comment by avaer** [comment]
I think some of these might be going down the wrong path like mushroom
house and pile of logs. To narrow this down, this is not an inhabited
forest with any industry or inhabitants to speak of, it is meant to be
lonely and confusing. Maybe it has some remnants and hints of something
that happened a while back like chopped stumps and remains of fences or
walls, or paths. But probably no houses or lumberjacks should be implied.

I think overall OOT got this theme right.

**Comment by avaer** [comment]
If there were some inhabitants implied it should be otherworldly in nature
like the skull kids rather than Hansel and Gretel.

**Comment by madjin** [comment]
![image](https://user-images.githubusercontent.com/32600939/122843991-acda0f80-d2ce-11eb-8876-531cfabb89dd.png)

**Comment by madjin** [comment]
![image](https://user-images.githubusercontent.com/32600939/123280149-8a1a4780-d4d6-11eb-988a-956a6ffa859b.png)

**Comment by avaer** [comment]
I wonder if it's too many tris when we instance a thousand of these. Reducing polys in things like the other wood is fine if we can pack more of these in.

Also can we make the trees straighter and taller? It is good that they are visible from the outside, but for artistic reasons I think they are better large and imposing, forming a dark structured canopy.

**Comment by madjin** [comment]
![image](https://user-images.githubusercontent.com/32600939/123458895-8f998f80-d5b3-11eb-8157-120f2ab98913.png)

**Comment by madjin** [comment]
![image](https://user-images.githubusercontent.com/32600939/123634836-c946e200-d7cf-11eb-8caf-54175004f340.png)
handpainting textures, will task on reducing triangles and making trees straighter / taller

**Comment by madjin** [comment]
DreamCo asked if other styles of trees might be better for this:

![image](https://user-images.githubusercontent.com/32600939/123679302-35d7d600-d7fc-11eb-8b3f-f025c835c9d1.png)
![image](https://user-images.githubusercontent.com/32600939/123679320-38d2c680-d7fc-11eb-9404-f2aeab24c166.png)

**Comment by madjin** [comment]
![image](https://user-images.githubusercontent.com/32600939/124290164-1fde5400-db21-11eb-89b8-fb70332b9f37.png)

![image](https://user-images.githubusercontent.com/32600939/124290180-2371db00-db21-11eb-98bc-06ffc6c418e0.png)

My feedback: The trees need more of a canopy so it feels enclosed in, also I'm not entirely sure about the water.

**Comment by madjin** [comment]
![image](https://user-images.githubusercontent.com/32600939/124625899-34d22480-de4c-11eb-92a4-3ccb9840db38.png)

**Comment by madjin** [comment]
Review / feedback on latest version:

Main thing: there's a lot of triangles, took a look at the blend file and noticed that the bushes are the densest clusters:

![image](https://user-images.githubusercontent.com/32600939/124785498-e50b6000-df14-11eb-84b7-e966d3f52299.png)

Each one is about ~1758 tris, we need to drastically reduce or remove these. In a single section it accounted for about 24k tris.

![image](https://user-images.githubusercontent.com/32600939/124785264-b7261b80-df14-11eb-81f8-714014802c14.png)

Next, as pretty as the colors look, we need to try out different color schemes and aesthetics.

![image](https://user-images.githubusercontent.com/32600939/124786478-bd68c780-df15-11eb-8c65-9b2e8ae0929a.png)

Example: [Black Forest](https://en.wikipedia.org/wiki/Black_Forest#/media/File:Green_winter.jpg), would look very cool with fog.

![image](https://user-images.githubusercontent.com/32600939/124786648-e4bf9480-df15-11eb-94de-fa22b83d9904.png)

The crystals don't really seem to make sense here. Although it would be interesting to have something that is mineable in the forest with a pickaxe, to gather some kind of material useful for crafting, not sure if crystals are part of the scope. Will remove in the blend.

![image](https://user-images.githubusercontent.com/32600939/124786767-fbfe8200-df15-11eb-99ad-2b0f858d1eee.png)

Feels like its getting close. Grant will crank out a couple new trim sheets tonight. Can then do another review, in-game testing, and optimization pass after.

**Comment by madjin** [comment]
Comparing canopies. We can get more coverage by widening the tree leaves or trying different tree types in general for less light / sky.

![image](https://user-images.githubusercontent.com/32600939/124788223-2997fb00-df17-11eb-93fe-d0b6d39c7a8f.png)

![image](https://user-images.githubusercontent.com/32600939/124788373-459b9c80-df17-11eb-8153-3aa41a7fec52.png)

![image](https://user-images.githubusercontent.com/32600939/124788489-6106a780-df17-11eb-8cf0-457b0c8532f6.png)

![image](https://user-images.githubusercontent.com/32600939/124788561-7380e100-df17-11eb-8e35-a39ecc4875c4.png)

**Comment by avaer** [comment]
Naturalizing the colors and adding god rays would help. In engine this would definitely have fog.

The crystals and mushrooms don't make too much sense, but some kind of vegetation does. Classes/clusters of grasses that can be replicated are probably the place to start.

The trees can go a lot higher and be lot more barren/wooden, except at the top.

Making any signs of inhbabitation old, destroyed, overgrown, and sparse would be closer in line with the theme.

**Comment by avaer** [comment]
The overall chunking and spacing is good, I think it comes down to details and textures.

**Comment by GaladWarder** [comment]
Did a retexture of this, new textures are provided in the Drive folder as well as packed into the Blend file: https://drive.google.com/drive/folders/1pyeQ7Ne9wcnvJr718CDIHgGYJyWbh3Dj?usp=sharing

**Comment by GaladWarder** [comment]
A few screenshots:

![Screenshot 2021-07-08 122010](https://user-images.githubusercontent.com/29031531/124861925-0d638f00-dfe7-11eb-8dea-15168a24f3f0.png)
![Screenshot 2021-07-08 122121](https://user-images.githubusercontent.com/29031531/124861932-0f2d5280-dfe7-11eb-98b5-5210aad403ae.png)
![Screenshot 2021-07-08 122309](https://user-images.githubusercontent.com/29031531/124862044-469bff00-dfe7-11eb-93ac-9d04057c1d0d.png)

Basically treated it like a color and mood pass - everything is darker and in general less saturated, with the exception of the greens - those are pushed a bit blbue and pumped up in vibrance to compensate for the lower sat elsewhere

**Comment by avaer** [comment]
I'm pretty sure we could just grab the zelda color values.

**Comment by avaer** [comment]
The walls/hills could also use a texturing pass to give the walls a bit of a cut gradients (but that  might be a UVing job first).

![image](https://user-images.githubusercontent.com/6926057/124874501-d022e100-df95-11eb-829b-eb0bdc4247c5.png)

**Comment by madjin** [comment]
Bushes and other plants have been optimized by lowering the poly count + rescattered all of them.

![image](https://user-images.githubusercontent.com/32600939/124933888-404f5800-dfd2-11eb-92bd-c1eaf960a84a.png)

The tree canopy has been changed to have less sky penetration into the forest

![image](https://user-images.githubusercontent.com/32600939/124934259-8dcbc500-dfd2-11eb-906c-30355aee20dc.png)

![image](https://user-images.githubusercontent.com/32600939/124934133-768cd780-dfd2-11eb-9eff-23c4de46cd94.png)

In other games like LOZ the canopies are a couple of planes. This works when the camera distance is limited so you can't really tell, but would be difficult to pull off in an open world scenario like in Fortnite where you can see from a distance and land on the Forest from the sky.

**Comment by avaer** [comment]
The vegetation is looking cool, though we will need to blend those textures better. Coming up with even more vegetation variations would probably be good.

**Comment by avaer** [comment]
The texture quality on the walls and floor seems to be suffering -- it looks lower than OOT. Whatever trim sheet/uvs technique is being used might need to change.

**Comment by avaer** [comment]
The trees still seem like the biggest problem. They are very friendly, healthy, and clean, but that isn't the vibe for a challenge like a Filter, which is what this is supposed to be. I'm more in favor of the thick planes, and making them much taller and more barren in the lower layers, like this reference.

![img](https://user-images.githubusercontent.com/32600939/124786648-e4bf9480-df15-11eb-94de-fa22b83d9904.png)

Having fog in screenshots might help to visualize.

We also don't need to be limited in the shapes of trees. There are other canopy styles. The trees do not need to be uniform and straight.

![download (3)](https://user-images.githubusercontent.com/6926057/124999172-95fc2280-e01b-11eb-8d2d-bdf6e32c4746.png)
![The Howling Forest, Alexander Ostrowski](https://user-images.githubusercontent.com/6926057/125001449-b1b5f780-e020-11eb-8fa0-d6dadf91555d.png)
![Desvre](https://user-images.githubusercontent.com/6926057/125001606-09546300-e021-11eb-8ece-90937ad790e5.png)

**Comment by avaer** [comment]
The rocks have a similar palette to the trees, which muddies things.

![image](https://user-images.githubusercontent.com/6926057/125002726-90a2d600-e023-11eb-8b2e-75fcc25aabb2.png)

I don't think we need such high poly rocks. The larger rocks seem like they should be more of a natural feature of the space rather than detailed stone slabs sticking out of the ground. The smaller rocks could be simplified and treated more as vegetation that is sprinkled around.

**Comment by madjin** [comment]
Rocks can be turned into ores with veins of something thats mineable
Screenshots should be closer to the reference images (for ex, enable fog in blender)
Ivy can be higher resolution (this asset is for trailer)
Lower density of signs, if there are signs can be destroyed
The textures need to be toned down to be darker / closer to what we had before:
![image](https://user-images.githubusercontent.com/32600939/125017833-e7b7a380-e041-11eb-982a-585fb084cd30.png)

**Comment by Vianvolaeus** [comment]
Updates from Dream today:

![image](https://user-images.githubusercontent.com/80061436/125105803-ca80e480-e0d6-11eb-88af-94dc2493c861.png)
![image](https://user-images.githubusercontent.com/80061436/125105812-cce33e80-e0d6-11eb-8a1b-519d0dfffe10.png)
![image](https://user-images.githubusercontent.com/80061436/125105827-d076c580-e0d6-11eb-89e0-1670ebdccb1c.png)
![image](https://user-images.githubusercontent.com/80061436/125105835-d2408900-e0d6-11eb-87a5-273243ce4b93.png)

**Comment by avaer** [comment]
The trees are very close in color to the dirt, and the textures are quite dark, making it hard to see.

**Comment by avaer** [comment]
The water is also green for some reason, rather than clear (which is what it would be at this shallow depth).

**Comment by madjin** [comment]
New tree variations

![image](https://user-images.githubusercontent.com/32600939/125478027-eac78bec-fc25-416a-a914-88804e97dbc9.png)

Side by side with previous iteration, the landscape now also uses a tilemap so its more easily updatable
![image](https://user-images.githubusercontent.com/32600939/125480916-baa6a41a-0fc3-48a6-874a-416e08960f04.png)
![image](https://user-images.githubusercontent.com/32600939/125479560-65f98eed-1863-48c7-9923-d1baa1c7fb67.png)

Here's a far shot for one of the modular pieces side by side a previous iteration
![image](https://user-images.githubusercontent.com/32600939/125480490-3bd114ca-13b6-4e35-beca-b096ca638962.png)

Water is now clearer and translucent 
![image](https://user-images.githubusercontent.com/32600939/125481605-bc9a7d38-203d-4ad6-aeba-babc8be5a3a9.png)

If greenlight, can finish the rest of the set (total 15 pieces modular land pieces).

**Comment by avaer** [comment]
The new shots are looking great, especially:

![img](https://user-images.githubusercontent.com/32600939/125479560-65f98eed-1863-48c7-9923-d1baa1c7fb67.png)
![image](https://user-images.githubusercontent.com/6926057/125521646-d40fd926-b47c-46b1-b877-cb46ea27e7bc.png)

It's not clear yet what the water looks like in context. For example, would it be possible to place an object at the bottom of a pool so that it could be noticed from above?

Otherwise I think this is a good level of detail for the chunks.

**Comment by madjin** [comment]
Here's a gem and wine bottle in the water for perspective
![image](https://user-images.githubusercontent.com/32600939/125524411-479dac80-20e0-4618-9ae9-3bd23d2b9af5.png)

**Comment by avaer** [comment]
From that shot, it seems like having ocean waves on the texture doesn’t make sense for this use, since it would either be still water or a flowing river; we should use a calmer water texture instead.

Is the intent to animate the texture or the water in any way?

**Comment by madjin** [comment]
Water is a whole thing. In the past I used gifs as water textures, then shaders.

- https://github.com/martinRenou/threejs-water
- https://threejs.org/examples/webgl_water.html
- https://threejs.org/examples/webgl_shaders_ocean.html

**Comment by avaer** [comment]
I’ve written many waters before, but this is more a question about the art direction/pipeline.

Ideally that is in the artist’s control, and it’s figured out early.

**Comment by avaer** [comment]
Fwiw I do think the water should be animated.

**Comment by madjin** [comment]
Few ideas:

- Editor.html where it's drag and drop to compose it into the scene (.glb) and export together
- Text editing the scene graph
  - 2 lines of code with janusweb to add cool water shader magic into a scene
  - Workflow: artist can position a placeholder asset and have shader_id="watershader" to enable effects
- Plane in the blender file that is named something like water_1
  - Blender plugin isn't being maintained so this idea is least realistic at the moment

**Comment by avaer** [comment]
That answer amounts to artists not making water; water is created exclusively by the programmer.

However if we are going this route then the water syntax/component decoration makes the most sense to me, including blender.

In the best case scenario we are copying hubs here, though potentially with our own actual water.

**Comment by Vianvolaeus** [comment]
Do we have any kind of UV scrolling possible in assets yet? I believe that's a commonly used effect for water and wouldn't detract from a regular texture flow for artists (much).

It's been referred to as a cheap toon shader effect elsewhere: https://virtualcast.jp/wiki/en/unity/shader/mtoonsample1

here's a small preview. It's a panned noise normal. 
![mtoonsea 1](https://user-images.githubusercontent.com/80061436/125539597-2f298fbe-a85d-47e1-9c90-990996107b17.gif)

**Comment by avaer** [comment]
Yes it should already be possible in the Hubs style coming from Blender.

**Comment by madjin** [comment]
Progress pics, full set will be ready for review soon.

![image](https://user-images.githubusercontent.com/32600939/125672952-29070e92-fb34-4bfd-95c3-275c6d7c96ae.png)

![image](https://user-images.githubusercontent.com/32600939/125672975-a7ff53a2-22f0-48e0-a0a7-599112e93347.png)

**Comment by madjin** [comment]
![image](https://user-images.githubusercontent.com/32600939/125808440-bc79a078-7445-4aea-b4da-f99ea0470973.png)

![image](https://user-images.githubusercontent.com/32600939/125808454-4b23d669-7038-425c-82aa-daaf951544f9.png)

10 modular pieces done, 5 more to go then optimization pass.

![Screenshot at 2021-07-15 11-06-50](https://user-images.githubusercontent.com/32600939/125811495-360b9aaa-2b4d-4b4c-b123-1a64898899dc.png)

**Comment by avaer** [comment]
This wouldn't prevent usage but:

1. From this angle it is looking quite generated rather than organic.
![image](https://user-images.githubusercontent.com/6926057/125872322-5d0860be-d615-4211-8b17-20f2db2e02ac.png)
1. Textures could be better UV'd and higher res, this is quite stretched.
![image](https://user-images.githubusercontent.com/6926057/125872430-d44ebfe7-f690-47a2-acce-bf506da5578b.png)

**Comment by madjin** [comment]
Think all the islands have been finished, checking the latest .blend that got uploaded to drive. Wanted to drop a couple screenshots first

![image](https://user-images.githubusercontent.com/32600939/125981944-7cb701b6-8a5a-49d2-b859-2b575755a9b1.png)

![image](https://user-images.githubusercontent.com/32600939/125981958-7e5baca6-69d3-4b66-892f-5d2a38e321bd.png)

I think it's going to look really good with fog, I could test in a game engine. Will take me a bit to setup.

**Comment by madjin** [comment]
All 15 pieces here

Tris: 2,268,227
96.1 MB

![blender_Kz6zxCbj78](https://user-images.githubusercontent.com/32600939/125984961-d33dcefe-7b95-433b-8917-02cd652b56bf.png)

![blender_NjZKHBE6uy](https://user-images.githubusercontent.com/32600939/125984965-5d762842-3c36-4cda-a9ae-c6bb962d168c.png)

19 textures
![image](https://user-images.githubusercontent.com/32600939/125986042-38fe9bc1-3ab2-42f3-8977-513f5b471031.png)

grass is 1024x1024, can try for higher res. UV is way more tedious to change. Few tweaks and I think its ready for next steps (renaming / optimizing pieces).

**Comment by avaer** [comment]
It looks good, but the texture quality and number of tris total is kind of alarming.

> I think it's going to look really good with fog, I could test in a game engine. Will take me a bit to setup.
Use Webaverse fog.

**Comment by madjin** [comment]
>  Use Webaverse fog.

Is there a code snippet for that somewhere?

**Comment by avaer** [comment]
I don't think any NFTs made it to master, but it's pretty easy to set up.

https://threejsfundamentals.org/threejs/lessons/threejs-fog.html

I actually am curious how it will look, but we can't really work with 2M tris for all of the pieces. Is it mostly in the trees?

**Comment by Vianvolaeus** [comment]
If you want me to take a look at this and figure out *why* it's such a high plyc. I can. Seems very high.

**Comment by Vianvolaeus** [comment]
Some of the tree variations are quite high polycount for what they are and their visibility to the player. A lot of the thinner branches are quite poly dense despite being really far from the player's view.

It *appears* these trees *weren't* batched during the creation of these different modular pieces.
If they are/were, then the offending pieces could be edited, and all instances of them would be changed in the scene at once. Since they are not, this becomes either a situation where the solution may be one of the following: 
a) multiple objects are manually edited individually
b) objects are manually replaced with an optimized version, instanced
c) aggressive decimation (akin to heavy LOD)

at 0.25 decimation in Blender applied to the entire scene, the polycount drops to 570k~.

Here is an example of tone of the modular pieces WITHOUT Decimation (top), and WITH Decimation (bottom)

From this perspective there is not much visual difference. If we're looking to push this ASAP and not use pieces for other individual projects then this might be a simple option. 


![image](https://user-images.githubusercontent.com/80061436/126003416-406db03c-74fb-47e6-9c8e-b865e559e836.png)

**Comment by avaer** [comment]
I think this will ultimately be a test case for LOD (which is part of the point!), but I think we should also trim the polygon count in the trees plus batch if we can.

Re batching, I think that can be done with `gltfpack`: `EXT_mesh_gpu_instancing`, and there is a THREE.js PR: https://github.com/mrdoob/three.js/issues/21937

**Comment by madjin** [comment]
The .blend needed to be modified so that the pipes can perfectly align with one another. Documenting here for future reference.

First select the pipes of one of the islands, then highlight the rest of the islands and hit ctrl+p (blender 279 keybind) and set parent.
![blender_0RUqL46bJ9](https://user-images.githubusercontent.com/32600939/126022190-f310e18d-628a-403f-b12a-12f92f7d21cb.png)


Next zero the transforms so its perfectly at center origin. Can hide and do the next island.
![blender_TpCKT1ohBY](https://user-images.githubusercontent.com/32600939/126022188-90814be7-83d9-4f11-967d-e77e44ef263c.png)

**Comment by avaer** [comment]
With the above make sure the collections/groups also have zeroed transforms, or it might affect the export.

The batching would be the last step in the pipeline by running through gltfpack.

**Comment by madjin** [comment]
All the collections are zeroed. I parented each island to the pipes, which are also zeroed.

**Comment by avaer** [comment]
Ok then here is the last step before it can be used: gotta consistently name each section based on its size and orientation.

For example if it has a top and right slot (N and E from birds eye), then it should have a name that ends with “_L” because the letter L implies that part. If all four sides are open then it would be “+”. It doesn’t matter the characters chosen but they need to be consistent. Finally we should encode the WxH of the piece in the name. They should already be following multiples of size, so a 2 in the W slot would mean it is a 2x fat chunk.

So a name could be:

1x1_-

Should around origin to reduce the amount of directions used. For example “-“ and “I” are isomorphic so can just be rotated to be one letter.

Also chunks names must be unique (blender/gltf thing).

**Comment by madjin** [comment]
Okay, I'll do a decimation pass then rename everything so we can begin testing. Later on we can replace the trees for instancing.

**Comment by madjin** [comment]
Going for human readability with LxW_NESW
So if its a 1x1 with the L shaped it will be 1x1_NE
If its everything it will be 1x1_NESW
![image](https://user-images.githubusercontent.com/32600939/126050575-58aedded-e187-44ed-b956-89a3d77f93c5.png)

**Comment by madjin** [comment]
## 1x1 checks

Modular 7 isn't perfectly aligned
![image](https://user-images.githubusercontent.com/32600939/126051249-b143bd83-7788-42f7-96bd-4b296a50c525.png)

Modular 11 isn't perfectly aligned
![blender_rMpX5lFET4](https://user-images.githubusercontent.com/32600939/126051293-58f82be1-f199-45ec-bffd-a784ddcf5a0a.png)


The pipes for all the 1x1 need to be perfectly aligned
![image](https://user-images.githubusercontent.com/32600939/126051291-52c45f96-7935-4777-8fd9-18313384335e.png)

## ETC
Modular 10 is a very different shape than the others
![image](https://user-images.githubusercontent.com/32600939/126051132-20c94452-cacf-471f-aa8c-a9dac1d69db8.png)

**Comment by madjin** [comment]
Hmm some pipes are shorter than others 
![image](https://user-images.githubusercontent.com/32600939/126052226-fc7c9627-2a9e-4cff-be4e-82226c1558a7.png)

**Comment by avaer** [comment]
Also we should disambiguate isomorphic naming. For example, NE is the same as NW since it is a rotation. Therefore only one of those suffixes should appear, and the algorithm will take care of rotations.

**Comment by madjin** [comment]
All the pipes have been aligned and renamed, going to get this ready for testing today!
